### PR TITLE
Abandon cURL

### PR DESCRIPTION
**Pull request by das-sein**
Part of the simplification project #1611 

Abandons the use of cURL entirely in favor of reqwest.

### CODE REVIEW

**Comment by das-sein** [comment]
Failure is due to Cargo.lock not being updated. Should I add that to the commit? @nrc @kinnison Wasn't sure if that's an okay thing to do on my side.

**Comment by dwijnand** [comment]
I'm not sure why `--locked` is used, but I think you should update it so CI validation passes, and then we can see if it's right?

**Comment by das-sein** [comment]
@kinnison @dwijnand Added Cargo.lock

**Comment by dwijnand** [comment]
The patch looks good to me.  The question is whether dropping curl in favour of reqwest is the right decision?

**Comment by briansmith** [comment]
> The patch looks good to me. The question is whether dumping curl in favour of reqwest is the right decision?

(I wish we weren't using words like "abandon" and "dumping" when talking about cURL since that's not very nice.)

I think replacing Curl with a Rust library is the right thing to do for both rhetorical (political) reasons and technical (security) reasons. So, I think the question isn't "whether" we should do something like this but what, if anything, do we need to do to make sure reqwest is a reasonable choice? The good thing about reqwest is that it is based on libraries that are widely used in production.

**Comment by das-sein** [comment]
@briansmith Sorry! I meant it in a totally non-emotional, literal way.

**Comment by jnicholls** [comment]
This looks good to me!

File: src/rustup-utils/src/notifications.rs [hunk 1]
```diff
@@ -62,12 +58,9 @@ impl<'a> Display for Notification<'a> {
...
             NoCanonicalPath(path) => write!(f, "could not canonicalize path: '{}'", path.display()),
             ResumingPartialDownload => write!(f, "resuming partial download"),
-            UsingCurl => write!(f, "downloading with curl"),
-            UsingReqwest => write!(f, "downloading with reqwest"),
-            UsingHyperDeprecated => f.write_str(
-                "RUSTUP_USE_HYPER environment variable is deprecated,\
-                 use RUSTUP_USE_REQWEST instead",
-            ),
+            UsingHyperDeprecated => {
+                f.write_str("RUSTUP_USE_HYPER environment variable is deprecated.")
```
Comment (seanmonstar) [inline src/rustup-utils/src/notifications.rs:62]: I don't have a strong feeling either way, but: should there be a warning about `RUSTUP_USE_REQWEST` being deprecated as well, as it is now the only option? If not, should the `RUSTUP_USE_HYPER` warning be removed?

Comment (das-sein) [reply src/rustup-utils/src/notifications.rs:62]: Good point, I think you're right that there should be some kind of warning. I'll add that in.

**Comment by brson** [comment]
It looks to me like prior to this patch curl was the default, so pretty much everybody has been using the curl backend. Switching to reqwest while also dropping the curl backend is risky. I recall the reqwest backend not working in some user scenarios (it's possible reqwest was even the default once and had to be switched back but I don't recall).

I'd suggest not doing this now, and instead turning on reqwest by default, releasing a new build, mentioning loudly in the release notes that the backend has changed, please report bugs, here's how to work around any problem (by requesting the curl backend). Wait a reasonable amount of time, then do this patch.

**Comment by brson** [comment]
> I'm not sure why `--locked` is used, but I think you should update it so CI validation passes, and then we can see if it's right?

It's so that Cargo.lock is kept updated, so that the release uses the same versions as the developers. Seems like a best practice to me; I don't see a downside.

**Comment by dwijnand** [comment]
Yeah, makes sense to me.  Just not used to it - in the cargo repo, even though it's also a binary tool, we're submoduled in, so we have no Cargo.lock.

**Comment by briansmith** [comment]
I agree it makes sense to have a way to switch back to using curl, e.g. via an environment variable, for a release.

**Comment by das-sein** [comment]
@brson Excellent points. I'll begin working on a PR that deprecates cURL without entirely removing it, and keep this PR open and up to date until it's safe to proceed.

**Comment by nrc** [comment]
Closing in favour of #1660

**Comment by briansmith** [comment]
@nrc #1660 is just a prerequisite to this, not a replacement for this.

**Comment by dwijnand** [comment]
@briansmith That's right.  The idea is let's go ahead and land #1660 instead, first, and remove cURL entirely after some time.

**Comment by lnicola** [comment]
Is it time to drop the cURL backend?

**Comment by jnicholls** [comment]
It's been 7 months. That is probably long enough.

**Comment by lnicola** [comment]
Filed #2034 for this.

**Pull request closed without merging**